app/services/chf/create_dzi_service.rb

```diff
@@ -0,0 +1,263 @@
+
+
+    attr_accessor :file_id, :checksum, :s3_bucket
+
+
+    def initialize(file_id, checksum: nil)
+      raise ArgumentError("file_id arg can not be empty") if file_id.blank?
+      @file_id = file_id
+      @checksum = checksum
+      @s3_bucket = self.class.s3_bucket!
+    end
+
+    # Downloads datastream and creates DZI files in WORKING_DIR,
+    # uploads them to S3.
+    #
+    # If lazy is true, will first check to see if the .dzi file already exists
+    # on S3, and if it does skip generation and upload. It doesn't check
+    # to make sure all tiles are correct, just dzi file exists.
+    def call(lazy: false)
+      if lazy && s3_bucket.object(dzi_file_name).exists?
+        return false
+      end
+
+      ensure_dirs
+
+      fetch_from_fedora!
+
+      create_dzi!
+
+      upload_to_s3!
+
+      return true
+    ensure
+      clean_up_tmp_files
+    end
+
+    protected
+
+    # Fetch Hydra::PCDM::File from fedora and save to WORKING_DIR,
+    # streaming for efficiency.
+    def fetch_from_fedora!
+      response = nil
+      # We hope this is the right way to get the actual uri to fetch cheaply?
+      uri = URI.parse(FileSet.translate_id_to_uri.call(file_id))
+
+      fedora_fetch_benchmark = Benchmark.measure do
+        response = Net::HTTP.start(uri.host, uri.port) do |http|
+          request = Net::HTTP::Get.new uri
+
+          if ActiveFedora.fedora.user || ActiveFedora.fedora.password
+            request.basic_auth(ActiveFedora.fedora.user, ActiveFedora.fedora.password)
+          end
+
+          http.request request do |response|
+            open local_original_file_path, 'wb' do |io|
+              response.read_body do |chunk|
+                io.write chunk
+              end
+            end
+          end
+        end
+      end
+      Rails.logger.debug("#{self.class.name}: fetch_from_fedora: #{fedora_fetch_benchmark}")
+      unless response.code == "200"
+        raise StandardError.new("Could not fetch file from fedora: #{uri}")
+      end
+    end
+
+    def create_dzi!
+      dzi_benchmark = Benchmark.measure do
+        # http://libvips.blogspot.com/2013/03/making-deepzoom-zoomify-and-google-maps.html
+        args = [vips_command, "dzsave", local_original_file_path, local_dzi_base_path, "--suffix", ".jpg[Q=#{jpeg_quality}]"]
+
+        if suppress_vips_stderr
+          args << { :err => "/dev/null" }
+        end
+
+        system(*args) or raise StandardError.new("vips dzsave failed")
+      end
+      Rails.logger.debug("#{self.class.name}: create_dzi: #{dzi_benchmark}")
+    end
+
+    # uploading to s3 is actually the slowest part if done serially.
+    # We apply a healthy dose of concurrency.
+    def upload_to_s3!
+      s_time = Time.now
+
+      # All the jpgs, which are in a _files/ dir, and subdirs of that.
+      futures = []
+      dir_path = Pathname.new(local_dzi_dir_path)
+      path_prefix_re = /\A#{Regexp.quote(WORKING_DIR.end_with?('/') ? WORKING_DIR : WORKING_DIR + '/')}/
```

Comment:
Added comment:       "We get back full paths from FS root from Dir.glob, need to change then to relative to working dir to get the S3 keys we want. "